### PR TITLE
[MM-46339] Handle user leaving channel with active call

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR implements the logic to handle the case of a user leaving the channel while in a call. Since they are effectively losing permissions to the call, we immediately disconnect the user and show a dialog so that they are aware of what happened (see screenshots).

Note: we may want to port the client-side changes to mobile to be consistent.

#### Screenshots

**User left**

image

**User was removed**

image

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-46339

### CODE REVIEW

File: webapp/src/components/call_error_modal.tsx [hunk 1]
```diff
@@ -125,6 +127,26 @@ export const CallErrorModal = (props: Props) => {
...
+    case userLeftChannelErr.message:
+        headerMsg = (
+            <span>{formatMessage({defaultMessage: 'You left the channel'})}</span>
+        );
+        msg = (
+            <span>
+                {formatMessage({defaultMessage: 'You were disconnected from the call as you left the channel.'})}
+            </span>
+        );
+        break;
```
Comment (streamer45) [inline webapp/src/components/call_error_modal.tsx:149]: @cwarnermm This is the part I am asking for editor review as we are adding a couple of new strings. Thanks!

File: server/plugin.go [hunk 2]
```diff
@@ -390,3 +390,40 @@ func (p *Plugin) MessageWillBeUpdated(c *plugin.Context, newPost, oldPost *model
...
+		p.LogWarn("UserHasLeftChannel: unexpected nil channel member")
+		return
+	}
+
+	state, err := p.kvGetChannelState(cm.ChannelId, false)
+	if err != nil {
+		p.LogError("UserHasLeftChannel: failed to get call state", "err", err.Error(), "channelID", cm.ChannelId)
+		return
+	} else if state == nil || state.Call == nil {
+		p.LogDebug("UserHasLeftChannel: no call ongoing", "channelID", cm.ChannelId)
```
Comment (cpoile) [inline server/plugin.go:405]: Do we need this? I'm worried it will spam the logs...

File: webapp/src/websocket_handlers.ts [hunk 3]
```diff
@@ -363,3 +368,14 @@ export function handleUserDismissedNotification(store: Store, ev: WebSocketMessa
...
 }
+
+export function handleUserRemovedFromChannel(store: Store, ev: WebSocketMessage<UserRemovedData>) {
+    const channelID = ev.data.channel_id || ev.broadcast.channel_id;
+    const currentUserID = getCurrentUserId(store.getState());
+    const removedUserID = ev.data.user_id || ev.broadcast.user_id;
+    const removerUserID = ev.data.remover_id;
+
+    if (removedUserID === currentUserID && channelID === channelIDForCurrentCall(store.getState())) {
+        getCallsClient()?.disconnect(removerUserID === currentUserID ? userLeftChannelErr : userRemovedFromChannelErr);
```
Comment (cpoile) [inline webapp/src/websocket_handlers.ts:379]: nice and simple

**Review by cpoile — APPROVED** [review]

File: server/plugin.go (hunk 2 above)
Comment (streamer45) [reply server/plugin.go:405]: That's the question, we need it when something suddenly breaks and we are clueless :)

I think that's what debug level logs are in my mind. It's a spam which you should be able to filter. It's more the times I had to debug something in production and wished I had these sorts of logs than the times I suffered through spam. But I agree with you, likely nothing too critical happening here so can probably be removed.

**Review by cwarnermm — COMMENTED** [review]
I propose the following copy updates visible to end users:

**You left the channel**
You have left the channel, and have been disconnected from the call.
[OK]

**You were removed from channel**
You have been removed from the channel, and have been disconnected from the call.
[OK]

**Review by esarafianou — APPROVED** [review]

**Review by abhijit-singh — COMMENTED** [review]
Thanks @streamer45, looks good! I was wondering if we can make the styling consistent with other modals. We're moving towards left aligned text and right aligned buttons in all the modals. Here's a figma link for reference - [Figma file](https://www.figma.com/file/G1xHnuLDxW4YPacSY4BW08/Calls%3A-UX-Nits?type=design&node-id=1528-138216&mode=design&t=9nsOMlE0IIFHl9os-4)
image

**Comment by streamer45** [comment]
Thanks @abhijit-singh! Should be improved now:

image

image

**Review by abhijit-singh — APPROVED** [review]
Looks good, thanks @streamer45!

**Pull request merged**